fx-src/FirefoxStorage.js

```diff
@@ -0,0 +1,227 @@
+
+  "updateData": `
+    UPDATE collection_data
+      SET record = :record
+        WHERE collection_name = :collection_name
+        AND record_id = :record_id`,
+
+  "deleteData": `
+    DELETE FROM collection_data
+      WHERE collection_name = :collection_name
+      AND record_id = :record_id`,
+
+  "saveLastModified": `
+    REPLACE INTO collection_metadata (collection_name, last_modified)
+      VALUES (:collection_name, :last_modified)`,
+
+  "getLastModified": `
+    SELECT last_modified
+      FROM collection_metadata
+        WHERE collection_name = :collection_name`,
+
+  "getRecord": `
+    SELECT record
+      FROM collection_data
+        WHERE collection_name = :collection_name
+        AND record_id = :record_id`,
+
+  "listRecords": `
+    SELECT record
+      FROM collection_data
+        WHERE collection_name = :collection_name`
+};
+
+const createStatements = ["createCollectionData",
+                          "createCollectionMetadata",
+                          "createCollectionDataRecordIdIndex"];
+
+export default class FirefoxAdapter extends BaseAdapter {
+  constructor(dbname) {
+    super();
+    this.dbname = dbname;
+
+    // make a promise to initialize
+    const self = this;
+    this.init = Task.spawn(function* () {
+      let connection;
+      try {
+        connection = yield self._doOpenConnection();
+        yield connection.executeTransaction(function* doSetup() {
+          let schema = yield (connection.getSchemaVersion());
+
+          if (schema == 0) {
+
+            for (let statementName of createStatements) {
+              yield (connection.execute(statements[statementName]));
+            }
+
+            yield (connection.setSchemaVersion(1));
+          } else if (schema != 1) {
+            throw new Error("Unknown database schema: " + schema);
+          }
+        });
+        self.initialized = true;
+      } finally {
+        if (connection) {
+          connection.close();
+        }
+      }
+    });
+  }
+
+  _doOpenConnection() {
+    const opts = { path: "kinto.sqlite", sharedMemoryCache: false }
+    return Sqlite.openConnection(opts);
+  }
+
+  _doTransaction(statement, params){
+    const self = this;
+    return Task.spawn(function* (){
+      // get a connection
+      let connection = yield self._doOpenConnection();
+
+      // ensure we're initialized
+      yield self.init;
+
+      //execute the transaction
+      try {
+        return yield connection.executeCached(statement, params);
+      }
+      finally {
+        // ensure the connection is closed
+        connection.close();
+      }
+    });
+  }
+
+  clear() {
+    const params = {collection_name: this.dbname};
+    return this._doTransaction(statements.clearData, params);
+  }
+
+  create(record) {
+    const params = {
+      collection_name: this.dbname,
+      record_id: record.id,
+      record: JSON.stringify(record)
+    };
+    return this._doTransaction(statements.createData, params).then(() => {
```

Comment:
> We're pretty much aware of that; the limitation is documented <http://kintojs.readthedocs.org/en/latest/limitations/#transactions>, we have an active issue <https://github.com/Kinto/kinto.js/issues/177> tagged for the very next release, we've discussed the implementation details and strategy on a lengthy thread <https://mail.mozilla.org/pipermail/kinto/2015-October/000019.html> on the Kinto development mailing-list… so I think we're covered here; just, it's a super tricky topic when it comes to IndexedDB, proper implementation should take some time to be done well and we badly needed to tag a first stable working version :)
> 

I’m actually not super concerned about transactions — we definitely want them, but not yet. I’m much more concerned about open/close semantics.